New buildsystem and merge of HPC-GAP

fingolfin opened the pull request:
DO NOT MERGE THIS VIA GITHUB WEB INTERFACE!

This pull request has been in the works for a long time. It implements a new build system, and also merges the hpcgap-default branch into the master in such a way that GAP and HPC-GAP coexist. Note that this branch was created with the help of a script in order to preserve history across both GAP and HPC-GAP source files. Moreover, this PR is not made against `master`, but instead against `hpc-merge-master` (without this trick, this PR would contain thousands of commits from HPC-GAP). As such, it should not be merged nor rebased via the github web interface; instead, if and when the time has come, I will run the aforementioned tool a last time in order to "rebase" this branch on the then-current master branch, and push it directly to update master. (Also, for a full review of the work in this PR, you may want to inspect `hpc-merge-master`, too).

This PR is crucial if we want to give HPC-GAP at least a fighting chance to survive and possibly become a regular part of GAP. Conversely, if we rejected it, then we would essentially reject HPC-GAP. 

Now on to a closer description of the content, and what is left to be done:

First off, the HPC-GAP part: This was done in accordance with the plan described on <https://github.com/gap-system/gap/wiki/Unifying-GAP4-and-HPC-GAP>. In particular, this PR does not naively merge master and hpcgap-default; instead, all files in the hpcgap-default branch were placed inside a new subdirectory `hpcgap`, without modifying any GAP library or kernel files. A normal build of GAP does not "see" any HPC-GAP files, so those who are not interested in HPC-GAP are not affected by it anymore than they are currently

Next a script was run to remove files from `hpcgap` which exist identically in the below the GAP root directory (e.g. `hpcgap/src/bool.c` was remove sinc `src/bool.c` has identical content). More files were later on synchronized and removed manually. The goal is to eventually have `hpcgap/src/` be empty. 

With this branch, building HPC-GAP roughly requires the following steps (assuming you are in a fresh clone of the GAP repository):
```
./autogen.sh   # generate configure via autoconf
git clone https://github.com/gap-system/ward  # get ward tool
cd ward && ./build.sh && cd .. # build ward
./configure --enable-hpcgap
make
```
(Note that in the future, ward will automatically be built, and perhaps also included with the GAP repository, so the above instructions would become even simpler).
As you can se,  there is no need to get a different branch, and with GAP 4.9, every GAP user could easily build and use HPC-GAP, if they wanted to.


Regarding the build system: The new buildsystem uses autoconf, libtool and GNU make. The old build system also used autoconf, but in a very non-standard way; the new build system is much more aligned with common practices, and should be easier adopted for future tasks.

In particular, we plan to add a working `make install` target, which should simplify packaging GAP, which in turn hopefully will result in GAP being available on more systems, in a more uniform way, and with shorter turnaround times. This is not yet in this PR, but will hopefully be added in the next couple weeks. In any case, the foundations for it are there.

Using GNU libtool makes it easier for us to build loadable modules (i.e. kernel extensions) portably and also enables us to build a shared library version of gap (dubbed `libgap`), which is highly interesting for sage, and likely also for the OSCAR project (and perhaps others). @markuspf has done quite some work on that front, but that will wait for a future PR.

A rough overview of the build system can be found in the file `README.buildsys.md`, and in comments in the various files that it consists of. This certainly could be improved (I should point out that the old build system was mostly undocumented). To any reviewers who find anything about it unclear, I'd appreciate feedback and questions -- if you find anything specific unclear, please say so in a comment, so that it can be explained and perhaps be better documented.

This PR is BIG. You may want to look at it commit-by-commit instead of browsing through the full diff. I did try to make clean and logical commits (except for the last few, which still need to be sorted and cleaned up). Also, note that there is still work to be done, though most (all?) of that could be don after merging this. For a list of issues and ideas, see <https://github.com/fingolfin/gap/issues> (the remaining issues there should of course be transferred to the GAP issue tracker, once this PR is merged).

Anyway, I think it is pointless to try to cover every detail of this PR now, so I will instead wait for people to ask questions.

CC @nthiery

fingolfin commented:
TODO: add an explanation of how the hpcgap mode works (include tricks, gap root hack). Will work on that when I am back in Germany

alex-konovalov commented:
In addition, this week I've created several Jenkins tests in addition to the Travis and Appveyor tests (also extended this week) to ensure that this pull requests gets the same level of testing that we use for the master branch. As a result, I confirm that:

-  All combinations of {install/standard/manuals/packagesload} x {Linux,macOS} x {32/64-bit builds} pass the test. 

- It compiles without warnings all combinations of  {32/64-bit builds} x {with/without readline} x {with/without enabling HPC-GAP} x {clang-xcode, gcc-cygwin, gcc-cygwin64, gcc48, gcc6-brew}

fingolfin commented:
I just updated this PR, rearranging some commits in order to hopefully make them more easily digestible and hence reviewable. I also added a section to `README.buildsys.md` which explains how building HPC-GAP is implemented.

Moreover, I was asked to point out that the new build system will not affect regular users at all (they can just run `./configure && make` as before); users (and probably also most developers) who build from git need to run on extra command once with the new setup  (namel `./autogen.sh`), though this is also a standard requirement for many other  projects. Only the handful (if any?) GAP developers who used the CONFIG feature of the old build system will have to adapt to using out-of-tree builds instead.

Finally, I would like to give an overview of the changes in this PR, in the hopes that this will help some people getting a grasp of this big PR. I am splitting the related commits into several phases:

## Phase I: Initial merge and scripted changes.

The changes in this phase are (re)generated by [a script](https://github.com/fingolfin/hpc-merge-files/blob/master/redo-branch.sh), so that one can conveniently rebase the whole hpc-merge branch on a newer version of master.

Note that these changes are not visible in the GitHub web view of this PR (in order to avoid several thousand extra commits from becoming visible). Rather, they are in the [hpc-merge-base branch](https://github.com/gap-system/gap/tree/hpc-merge-base). Summary of changes:

* 6a7aab05107afd7340c8684207c23a70a2cdbe16 Move hpcgap files into hpcgap subdirectory (before merge)
* 57592e09605059d81d12d9534878d7b7afaed77d Merge hpcgap-default into hpc-merge
* https://github.com/gap-system/gap/compare/57592e0...fingolfin:hpc-merge-base Remove old build system, rewrite #include directives, remove lots of redundant files from `hpcgap/`

## Phase II: Preparations

These are the changes up to just before the addition of the new build system:
https://github.com/gap-system/gap/compare/hpc-merge-base...d16d1c2

* Clean up the `extern/` and `hpcgap/extern/` directories.
* Some generated files are renamed:
  - `src/c_filt1.c` -> `src/c_filter1.c`
  - `src/c_meths1.c` -> `src/c_methsel1.c`
* Remove ZeroMQ support (now in a package)
* Add GAP compiler to emit `#include <src/compiled.h>`
* Various other minor tweaks

# Phase III: New build system

* 34256e3accd493e8bf67bf78fc1d85990c243a75 Add new build system -- this is *THE* central commit of this PR
* 5c80ce3be62d51056b89b41c16ec50e50e9523f7 Add config.guess, config.sub and install-sh to repository
* b5b4d7bb7cc77398f3ebed83dc0c65ae921f5ee1 Add script to update config.{guess,sub} 
* 8837fb88e3de3ae57262591ceadf2ed67d57f07a bin/BuildPackages.sh: replace HACK for detecting 32bit builds with another HACK
* bcc9d2b959908e531d3c969954629790d76f89ea Makefile.rules: add targets for the GAP test suite 

# Phase IV: CI tests, HPC-GAP syncs and other stuff

Commits in this phase: https://github.com/gap-system/gap/compare/bcc9d2b959908e531d3c969954629790d76f89ea...fingolfin:hpc-merge

These commits roughly fall into three groups:
* enable or improve Travis and AppVeyor integration
* sync changes between GAP and HPC-GAP (mostly focused on changes to HPC-GAP, only very minor things are applied to the GAP side, like whitespace fixes)
* a few minor GAP kernel tweaks (e.g. 5eb26d0799d5ec8f538bbc30f1b7261568cfea9b) that came up while I worked on this and that I did not (yet???) backport to master, in the hopes that we can just merge this.

ChrisJefferson commented on README.buildsys.md:
```diff
@@ -0,0 +1,207 @@
+In order to work on the build system, you need at least the following:
+
+* GNU autoconf (we recommend 2.69 or later)
+* GNU libtool (we recommend 2.4.2 or later)
```
also automake

fingolfin replied:
Hmm, we don't actually use automake?! ... [ /me wanders off to try it out ]
Ahhh, turns out autoreconf does seem to use it indirectly, by calling aclocal, which is part of automake. But in truth, we don't need automake. I'll see what I can do to resolve this.

markuspf replied:
Does GMP use automake?

fingolfin replied:
GMP, Boehm GC and libatomic_ops all use automake (note the presence of `Makefile.am` files). But this is irrelevant, as we don't need to run (the equivalent of) `autogen.sh` for them.

ChrisJefferson commented on README.buildsys.md:
```diff
@@ -0,0 +1,207 @@
+modified so that for every root directory `FOO` that gets added, we first add
+`FOO/hpcgap` to the list of root directories. 
+
+5. Ward (a tool for scanning our C sources and inserting guard statements into it) is integrated as follows: Normally, `src/FOO.c` gets compiled to `obj/FOO.o`. With ward in the mix, `src/FOO.c` or (if it exists) `hpcgap/src/FOO.c` is first turned into `gen/FOO.c` by ward, and then the compiler turns the latter into `obj/FOO.o`. For each conversion, dependencies are tracked via `.../.deps/FOO.d`  ...
```
this paragraph isn't wrapped

stevelinton commented:
This builds for me but when run I get a `#W` Static module lib/filter1.g has CRC mismatch, ignoring`  Also `etc/docomp` seems to have disappeared, so I can't rebuild `c_filter1.c`.

fingolfin commented:
@stevelinton That's weird, for me the CRCs match... Was this with plain GAP or HPC-GAP.

The `etc/docomp` script was replaced by `make docomp`, reflecting the fact that it requires a working GAP, plus that there can be differences between the GAP and HPC-GAP version.

stevelinton commented:
Plain GAP. I’ll check more carefully if I have any modified files anywhere or anything.

	Steve

> On 22 Mar 2017, at 14:37, Max Horn <notifications@github.com> wrote:
> 
> @stevelinton That's weird, for me the CRCs match... Was this with plain GAP or HPC-GAP.
> 
> The etc/docomp script was replaced by make docomp, reflecting the fact that it requires a working GAP, plus that there can be differences between the GAP and HPC-GAP version.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 
> {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/gap-system/gap","title":"gap-system/gap","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/gap-system/gap"}},"updates":{"snippets":[{"icon":"PERSON","message":"@fingolfin in #1193: @stevelinton That's weird, for me the CRCs match... Was this with plain GAP or HPC-GAP.\r\n\r\nThe `etc/docomp` script was replaced by `make docomp`, reflecting the fact that it requires a working GAP, plus that there can be differences between the GAP and HPC-GAP version."}],"action":{"name":"View Pull Request","url":"https://github.com/gap-system/gap/pull/1193#issuecomment-288418903"}}}

fingolfin commented:
Is this with the latest revision of this? Make sure you have the right ref (as I write this, tip of this branch is `059cb938770cb0e4db7a9e5fe79464e3e1ef846a`, though I'll post some more work soon ;-)

fingolfin commented:
For the record, I revised and extend this work some more; but since people might be reviewing and/or testing this PR, I have not (yet) pushed it here. Instead, this is for now on a separate branch, see https://github.com/fingolfin/gap/tree/hpc-merge-full resp. https://github.com/gap-system/gap/compare/hpc-merge-base...fingolfin:hpc-merge-full

Besides squashing some fixup commits and some additional general improvements, that version does more HPC-GAP syncing, and also cleans up existing sync work: Now, I made sure that when syncing files, we first commit a synced version to `hpcgap/`; files which are fully synced are then only removed in a later commit. This has the advantage that it becomes much easier to review these commits for correctness, as it is immediately clear what changes were synced; and the commits which "remove duplicates" now don't need to be checked, as they are created by a script that really only removes files that are identical. This is useful even after a merge: If I screwed up syncing any files, one can now find out by inspecting the history relatively easily.

fingolfin commented:
A revised version of this has now been merged :-).

mohamed-barakat commented:
Congratulations 😊 (With no room for mistake this time 😉)
